internal/transport/http2_server.go

```diff
@@ -371,32 +371,43 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 			timeoutSet = true
 			var err error
 			if timeout, err = decodeTimeout(hf.Value); err != nil {
-				headerError = true
+				headerError = err
 			}
 		default:
 			if isReservedHeader(hf.Name) && !isWhitelistedHeader(hf.Name) {
 				break
 			}
 			v, err := decodeMetadataHeader(hf.Name, hf.Value)
 			if err != nil {
-				headerError = true
+				headerError = err
 				logger.Warningf("Failed to decode metadata header (%q, %q): %v", hf.Name, hf.Value, err)
 				break
 			}
 			mdata[hf.Name] = append(mdata[hf.Name], v)
 		}
 	}
 
-	if !isGRPC || headerError {
-		t.controlBuf.put(&cleanupStream{
+	if !isGRPC {
+		t.controlBuf.put(&headerFrame{
```

Comment:
Does this trigger a RST_STREAM afterwards?  It should, but we can skip it if the header we are processing has END_STREAM set on it.

Reply:
I don't see the put triggering an RST_STREAM but that may be that I'm unfamiliar with it. Do you mean we can skip the put if the stream is ended?